gateway-spi/src/main/java/org/apache/knox/gateway/util/GroupBasedImpersonationProvider.java

```diff
@@ -0,0 +1,263 @@
+        Map<String, String> allMatchKeys =
+                getConf().getValByRegex(usersGroupsRegEx);
+
+        for (Map.Entry<String, String> entry : allMatchKeys.entrySet()) {
+            //aclKey = hadoop.proxygroup.[VIRTUAL_GROUP]
+            String aclKey = getAclKey(entry.getKey());
+
+            if (!proxyGroupsAcls.containsKey(aclKey)) {
+                proxyGroupsAcls.put(aclKey, new AccessControlList(
+                        allMatchKeys.get(aclKey + CONF_USERS),
+                        allMatchKeys.get(aclKey + CONF_GROUPS)));
+            }
+        }
+
+        // get hosts per proxygroup
+        allMatchKeys = getConf().getValByRegex(hostsRegEx);
+        for (Map.Entry<String, String> entry : allMatchKeys.entrySet()) {
+            groupProxyHosts.put(entry.getKey(),
+                    new MachineList(entry.getValue()));
+        }
+    }
+
+    private String getAclKey(String key) {
+        int endIndex = key.lastIndexOf('.');
+        if (endIndex != -1) {
+            return key.substring(0, endIndex);
+        }
+        return key;
+    }
+
+    /**
+     * Authorization based on user and group impersonation policies.
+     *
+     * @param user the user information attempting the operation, which includes the real
+     *             user and the effective impersonated user.
+     * @param remoteAddress the remote address from which the user is connecting.
+     * @throws AuthorizationException if the user is not authorized based on the
+     *                                configured impersonation and group policies.
+     */
+    @Override
+    public void authorize(UserGroupInformation user, InetAddress remoteAddress) throws AuthorizationException {
+        authorize(user, remoteAddress, Collections.emptyList());
+    }
+
+    /**
+     * Authorization based on groups that are provided as a function argument
+     *
+     * @param user the user information attempting the operation, which includes the real
+     *             user and the effective impersonated user.
+     * @param groups the list of groups to check for authorization.
+     * @param remoteAddress the remote address from which the user is connecting.
+     * @throws AuthorizationException if the user is not authorized based on the
+     *                                configured impersonation and group policies.
+     */
+    public void authorize(UserGroupInformation user, InetAddress remoteAddress, List<String> groups) throws AuthorizationException {
+
+        /* Proxy user configuration takes precedence over proxy group configuration. */
+        if (doesProxyUserConfigExist) {
+            try{
+                /* check for proxy user authorization */
+                super.authorize(user, remoteAddress);
+            } catch (final AuthorizationException e) {
+                /*
+                 * Log and try group based impersonation.
+                 * Since this provider is for groups no need to check if
```

Comment:
Is this comment true? We only employ this implementation if there is group-level proxy config? That seems wrong to me.

Reply:
The main reason was separation of functionality. I did not want the existing functionality to be affected by new code. So, unless users configure proxy group `GroupBasedImpersonationProvider` will not be engaged. This was my thinking.